// @knodes/typedoc-plugin-pages/dist/models/reflections/a-node-reflection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ANodeReflection = void 0;
const typedoc_1 = require("typedoc");
const reflection_kind_1 = require("./reflection-kind");
class ANodeReflection extends typedoc_1.ContainerReflection {
    get depth() {
        return this.parent instanceof ANodeReflection ? this.parent.depth + 1 : 0;
    }
    get isModuleAppendix() {
        return this.module === this.parent && this.name === this.module.name;
    }
    constructor(name, kind, module, parent = module) {
        super(name, kind | (module === parent ? reflection_kind_1.PagesPluginReflectionKind.ROOT : 0), parent);
        this.module = module;
        this.parent = parent;
    }
    /**
     * Iterate over all children & childrenNodes of this node reflection.
     *
     * @param callback - The function to invoke on each child.
     */
    traverse(callback) {
        var _a, _b;
        super.traverse(callback);
        for (const c of (_b = (_a = this.childrenNodes) === null || _a === void 0 ? void 0 : _a.slice()) !== null && _b !== void 0 ? _b : []) {
            if (callback(c, typedoc_1.TraverseProperty.Children) === false) {
                return;
            }
        }
    }
}
exports.ANodeReflection = ANodeReflection;
