// @knodes/typedoc-plugin-pages/dist/models/reflections/page-reflection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PageReflection = void 0;
const fs_1 = require("fs");
const typedoc_1 = require("typedoc");
const typedoc_pluginutils_1 = require("@knodes/typedoc-pluginutils");
const path_1 = require("@knodes/typedoc-pluginutils/path");
const a_node_reflection_1 = require("./a-node-reflection");
const reflection_kind_1 = require("./reflection-kind");
class PageReflection extends a_node_reflection_1.ANodeReflection {
    constructor(name, module, parent, sourceFilePath, nodeVirtualPath, url) {
        super(name, reflection_kind_1.PagesPluginReflectionKind.PAGE, module, parent);
        this.sourceFilePath = sourceFilePath;
        this.url = url;
        this.content = typedoc_pluginutils_1.miscUtils.catchWrap(() => (0, fs_1.readFileSync)(sourceFilePath, 'utf-8'), `Could not read ${(0, path_1.relative)(process.cwd(), sourceFilePath)}`);
        this.sources = [
            new typedoc_1.SourceReference(sourceFilePath, 1, 1),
        ];
        this.comment = new typedoc_1.Comment([{ kind: 'text', text: this.content }]);
        this.relevanceBoost = 5;
        const namedPathRoot = this.module instanceof typedoc_1.ProjectReflection ? '~~' : `~/${this.module.name}`;
        this.namedPath = (0, typedoc_1.normalizePath)(`${namedPathRoot}/${nodeVirtualPath}`.toLowerCase());
    }
}
exports.PageReflection = PageReflection;
