// @knodes/typedoc-plugin-pages/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ANodeReflection = exports.PagesPluginReflectionKind = exports.MenuReflection = exports.PageReflection = exports.EInvalidPageLinkHandling = exports.load = void 0;
var load_1 = require("./load");
Object.defineProperty(exports, "load", { enumerable: true, get: function () { return load_1.load; } });
var options_1 = require("./options");
Object.defineProperty(exports, "EInvalidPageLinkHandling", { enumerable: true, get: function () { return options_1.EInvalidPageLinkHandling; } });
var reflections_1 = require("./models/reflections");
Object.defineProperty(exports, "PageReflection", { enumerable: true, get: function () { return reflections_1.PageReflection; } });
Object.defineProperty(exports, "MenuReflection", { enumerable: true, get: function () { return reflections_1.MenuReflection; } });
Object.defineProperty(exports, "PagesPluginReflectionKind", { enumerable: true, get: function () { return reflections_1.PagesPluginReflectionKind; } });
Object.defineProperty(exports, "ANodeReflection", { enumerable: true, get: function () { return reflections_1.ANodeReflection; } });
