// @knodes/typedoc-pluginutils/dist/utils/path.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.extname = exports.basename = exports.parse = exports.isAbsolute = exports.normalize = exports.dirname = exports.relative = exports.join = exports.resolve = void 0;
// eslint-disable-next-line no-restricted-imports -- Re-export normalized
const path_1 = require("path");
Object.defineProperty(exports, "basename", { enumerable: true, get: function () { return path_1.basename; } });
Object.defineProperty(exports, "extname", { enumerable: true, get: function () { return path_1.extname; } });
Object.defineProperty(exports, "isAbsolute", { enumerable: true, get: function () { return path_1.isAbsolute; } });
Object.defineProperty(exports, "parse", { enumerable: true, get: function () { return path_1.parse; } });
const lodash_1 = require("lodash");
const typedoc_1 = require("typedoc");
const resolve = (from, ...to) => (0, typedoc_1.normalizePath)((0, path_1.resolve)(from, ...to.filter(lodash_1.isString)));
exports.resolve = resolve;
const join = (...paths) => (0, typedoc_1.normalizePath)((0, path_1.join)(...paths.filter(lodash_1.isString)));
exports.join = join;
const relative = (from, to) => (0, typedoc_1.normalizePath)((0, path_1.relative)(from, to));
exports.relative = relative;
const dirname = (path) => (0, typedoc_1.normalizePath)((0, path_1.dirname)(path));
exports.dirname = dirname;
const normalize = (path) => (0, typedoc_1.normalizePath)((0, path_1.normalize)(path));
exports.normalize = normalize;
