// @knodes/typedoc-plugin-pages/dist/options/build.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildOptions = void 0;
const assert_1 = __importStar(require("assert"));
const lodash_1 = require("lodash");
const typedoc_1 = require("typedoc");
const typedoc_pluginutils_1 = require("@knodes/typedoc-pluginutils");
const types_1 = require("./types");
const wrapPageError = (path, index) => (err) => {
    var _a;
    if (err instanceof assert_1.AssertionError) {
        const pathStr = path.length > 0 ? ` in ${path.map(p => JSON.stringify(p)).join(' ⇒ ')}` : '';
        return new Error(`Invalid page${pathStr} @ index ${index}: ${(_a = err.message) !== null && _a !== void 0 ? _a : err}`, { cause: err });
    }
    else {
        return err;
    }
};
const pageKeys = ['children', 'childrenDir', 'childrenOutputDir', 'childrenSourceDir', 'output', 'source', 'name'];
const checkPageFactory = (allowedKeys) => (plugin, page, path) => {
    (0, assert_1.default)(page && (0, lodash_1.isObject)(page), 'Page should be an object');
    const _page = page;
    if ('title' in _page && !('name' in _page)) {
        _page.name = _page.title;
        delete _page.title;
        plugin.logger.warn(`Page ${[...path, _page.name].map(p => `"${p}"`).join(' ⇒ ')} is using deprecated "title" property. Use "name" instead.`);
    }
    (0, assert_1.default)('name' in _page && (0, lodash_1.isString)(_page.name), 'Page should have a name');
    const extraProps = (0, lodash_1.difference)(Object.keys(_page), allowedKeys);
    assert_1.default.equal(extraProps.length, 0, `Page ${[...path, _page.name].map(p => `"${p}"`).join(' ⇒ ')} have extra properties ${JSON.stringify(extraProps)}`);
    if ('children' in _page && !(0, lodash_1.isNil)(_page.children)) {
        (0, assert_1.default)((0, lodash_1.isArray)(_page.children), 'Page children should be an array');
        const thisPath = [...path, _page.name];
        _page.children.forEach((c, i) => typedoc_pluginutils_1.miscUtils.catchWrap(() => checkPage(plugin, c, thisPath), wrapPageError(thisPath, i)));
    }
};
const checkPage = checkPageFactory(pageKeys);
const checkRootPage = checkPageFactory([...pageKeys, 'moduleRoot']);
const buildOptions = (plugin) => typedoc_pluginutils_1.OptionGroup.factory(plugin)
    .add('enablePageLinks', {
    help: 'Whether or not @page and @pagelink tags should be parsed.',
    type: typedoc_1.ParameterType.Boolean,
    defaultValue: true,
})
    .add('enableSearch', {
    help: 'Whether or not the pages should be added to the search index.',
    type: typedoc_1.ParameterType.Boolean,
    defaultValue: true,
})
    .add('invalidPageLinkHandling', {
    help: 'The kind of error to throw in case of an invalid page link.',
    type: typedoc_1.ParameterType.Map,
    map: types_1.EInvalidPageLinkHandling,
    defaultValue: types_1.EInvalidPageLinkHandling.LOG_ERROR,
})
    .add('pages', {
    help: 'Actual pages definitions.',
    type: typedoc_1.ParameterType.Mixed,
    validate: v => {
        if (v) {
            (0, assert_1.default)((0, lodash_1.isArray)(v), 'Pages should be an array');
            if (!v.length) {
                return;
            }
            v.forEach((p, i) => typedoc_pluginutils_1.miscUtils.catchWrap(() => checkRootPage(plugin, p, []), wrapPageError([], i)));
            const rootFlags = (0, lodash_1.groupBy)(v, p => !!p.moduleRoot);
            assert_1.default.equal(Object.keys(rootFlags).length, 1, 'Every root pages should set `moduleRoot` to true, or none');
            if (rootFlags.true) {
                assert_1.default.equal((0, lodash_1.uniq)(v.map(p => p.name)).length, v.length, 'Every root pages should have a different name');
            }
        }
    },
}, v => {
    v = v !== null && v !== void 0 ? v : [];
    return v;
})
    .add('output', {
    help: 'Output directory where your pages will be rendered. This must be a relative path.',
    type: typedoc_1.ParameterType.String,
    defaultValue: 'pages',
}, v => plugin.relativeToRoot(v))
    .add('source', {
    help: 'Root directory where all page source files live.',
    type: typedoc_1.ParameterType.String,
    defaultValue: 'pages',
})
    .add('logLevel', {
    help: 'The plugin log level.',
    type: typedoc_1.ParameterType.Map,
    map: typedoc_1.LogLevel,
    defaultValue: plugin.application.logger.level,
})
    .add('excludeMarkdownTags', {
    help: 'A list of markdown captures to omit. Should have the form `{@....}`.',
    type: typedoc_1.ParameterType.Array,
    validate: patterns => patterns === null || patterns === void 0 ? void 0 : patterns.forEach(p => assert_1.default.match(p, /^\{@.*\}$/, `Pattern ${JSON.stringify(p)} should match "{@...}"`)),
}, v => v !== null && v !== void 0 ? v : [])
    .build();
exports.buildOptions = buildOptions;
